lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/FunctionAccessControlRule.java

```diff
@@ -0,0 +1,144 @@
+import static io.trino.plugin.base.security.FunctionAccessControlRule.FunctionPrivilege.GRANT_EXECUTE;
+import static io.trino.spi.function.FunctionKind.AGGREGATE;
+import static io.trino.spi.function.FunctionKind.SCALAR;
+import static io.trino.spi.function.FunctionKind.TABLE;
+import static io.trino.spi.function.FunctionKind.WINDOW;
+import static java.util.Objects.requireNonNull;
+
+public class FunctionAccessControlRule
+{
+    public static final FunctionAccessControlRule ALLOW_ALL = new FunctionAccessControlRule(
+            ImmutableSet.copyOf(FunctionPrivilege.values()),
+            Optional.empty(),
+            Optional.empty(),
+            Optional.empty(),
+            Optional.empty(),
+            ImmutableSet.copyOf(FunctionKind.values()),
+            Optional.empty());
+
+    private final Set<FunctionPrivilege> privileges;
+    private final Optional<Pattern> userRegex;
+    private final Optional<Pattern> roleRegex;
+    private final Optional<Pattern> groupRegex;
+    private final Optional<Pattern> schemaRegex;
+    private final Set<FunctionKind> functionKinds;
+    private final Optional<Pattern> functionRegex;
+
+    @JsonCreator
+    public FunctionAccessControlRule(
+            @JsonProperty("privileges") Set<FunctionPrivilege> privileges,
+            @JsonProperty("user") Optional<Pattern> userRegex,
+            @JsonProperty("role") Optional<Pattern> roleRegex,
+            @JsonProperty("group") Optional<Pattern> groupRegex,
+            @JsonProperty("schema") Optional<Pattern> schemaRegex,
+            @JsonProperty("function_kinds") @JsonAlias("functionKinds") Set<FunctionKind> functionKinds,
+            @JsonProperty("function") Optional<Pattern> functionRegex)
+    {
+        this.privileges = ImmutableSet.copyOf(requireNonNull(privileges, "privileges is null"));
+        this.userRegex = requireNonNull(userRegex, "userRegex is null");
+        this.roleRegex = requireNonNull(roleRegex, "roleRegex is null");
+        this.groupRegex = requireNonNull(groupRegex, "groupRegex is null");
+        this.schemaRegex = requireNonNull(schemaRegex, "schemaRegex is null");
+        this.functionKinds = requireNonNull(functionKinds, "functionKinds is null");
+        this.functionRegex = requireNonNull(functionRegex, "functionRegex is null");
+    }
+
+    public boolean matches(String user, Set<String> roles, Set<String> groups, String functionName)
+    {
+        return (functionKinds.contains(SCALAR) || functionKinds.contains(AGGREGATE) || functionKinds.contains(WINDOW)) &&
+                userRegex.map(regex -> regex.matcher(user).matches()).orElse(true) &&
+                roleRegex.map(regex -> roles.stream().anyMatch(role -> regex.matcher(role).matches())).orElse(true) &&
+                groupRegex.map(regex -> groups.stream().anyMatch(group -> regex.matcher(group).matches())).orElse(true) &&
+                functionRegex.map(regex -> regex.matcher(functionName).matches()).orElse(true);
```

Comment:
I think you need to make sure that schema regex is empty here. Please add a test for this.

Reply:
https://github.com/trinodb/trino/pull/13713#discussion_r998171741